include/nbl/video/IGPUCommandPool.h

```diff
@@ -4,38 +4,835 @@
 
 #include "nbl/video/decl/IBackendObject.h"
-
+#include "nbl/video/IGPUPipelineLayout.h"
 
 namespace nbl::video
 {
+class IGPUCommandBuffer;
 
 class NBL_API IGPUCommandPool : public core::IReferenceCounted, public IBackendObject
 {
+public:
+    static inline constexpr uint32_t COMMAND_SEGMENT_SIZE = 128u << 10u;
+
+    enum E_CREATE_FLAGS : uint32_t
+    {
+        ECF_NONE = 0x00,
+        ECF_TRANSIENT_BIT = 0x01,
+        ECF_RESET_COMMAND_BUFFER_BIT = 0x02,
+        ECF_PROTECTED_BIT = 0x04
+    };
+
+private:
+    static inline constexpr uint32_t COMMAND_ALIGNMENT = 64u;
+
+    static inline constexpr uint32_t COMMAND_SEGMENT_ALIGNMENT = 64u;
+
+    static inline constexpr uint32_t MAX_COMMAND_SEGMENT_BLOCK_COUNT = 16u;
+    static inline constexpr uint32_t COMMAND_SEGMENTS_PER_BLOCK = 256u;
+    static inline constexpr uint32_t MIN_POOL_ALLOC_SIZE = COMMAND_SEGMENT_SIZE;
+
+public:
+    class CCommandSegment;
+
+    class alignas(COMMAND_ALIGNMENT) ICommand
+    {
+    public:
+        virtual ~ICommand() {}
+
+        // static void* operator new(std::size_t size) = delete;
+        static void* operator new[](std::size_t size) = delete;
+        // static void* operator new(std::size_t size, std::align_val_t al) = delete;
+        static void* operator new[](std::size_t size, std::align_val_t al) = delete;
+
+        // static void operator delete  (void* ptr) = delete;
+        static void operator delete[](void* ptr) = delete;
+        static void operator delete  (void* ptr, std::align_val_t al) = delete;
+        static void operator delete[](void* ptr, std::align_val_t al) = delete;
+        static void operator delete  (void* ptr, std::size_t sz) = delete;
+        static void operator delete[](void* ptr, std::size_t sz) = delete;
+        static void operator delete  (void* ptr, std::size_t sz, std::align_val_t al) = delete;
+        static void operator delete[](void* ptr, std::size_t sz, std::align_val_t al) = delete;
+
+        inline uint32_t getSize() const { return m_size; }
+
+    protected:
+        ICommand(uint32_t size) : m_size(size) {}
+
+    private:
+        friend CCommandSegment;
+
+        const uint32_t m_size;
+    };
+
+    template <class CRTP>
+    class NBL_FORCE_EBO IFixedSizeCommand : public IGPUCommandPool::ICommand
+    {
     public:
-        enum E_CREATE_FLAGS : uint32_t
+        template <typename... Args>
+        static uint32_t calc_size(const Args&...)
+        {
+            return sizeof(CRTP);
+        }
+
+    protected:
+        IFixedSizeCommand() : ICommand(calc_size()) {}
+    };
+
+    class alignas(COMMAND_SEGMENT_ALIGNMENT) CCommandSegment
+    {
+    private:
+        struct header_t
         {
-            ECF_NONE = 0x00,
-            ECF_TRANSIENT_BIT = 0x01,
-            ECF_RESET_COMMAND_BUFFER_BIT = 0x02,
-            ECF_PROTECTED_BIT = 0x04
+            core::LinearAddressAllocator<uint32_t> m_commandAllocator;
+            CCommandSegment* m_next;
```

Comment:
We could make this data structure a 2D woven net like so
```mermaid
stateDiagram-v2
    mainHead : This Tape's HEAD
    headA : Some other HEAD
    headB : Some other HEAD
    interiorSegment : Some interior segment
    interiorSegmentA : Some interior segment
    interiorSegmentB : Some interior segment
    tail : Tail
    headA --> mainHead
    mainHead --> headA
    headB --> mainHead
    mainHead --> headB
    mainHead --> interiorSegment
    interiorSegment --> tail
    headA --> interiorSegmentA
    headB --> interiorSegmentB
```

Reply:
```cpp
        struct header_t
        {
            core::LinearAddressAllocator<uint32_t> m_commandAllocator;
            CCommandSegment* m_next;
            // the following are only set for tape HEAD segments
            CCommandSegment* m_prevHead = nullptr;
            CCommandSegment* m_nextHead = nullptr;
        } header;
```

Your `CCommandSegmentPool::emplace` and `CCommandSegmentPool::delete` would need to be slightly patched up to take into account that they're adding to the end or removing from the middle of a doubly linked list.
(assuming you implement https://github.com/Devsh-Graphics-Programming/Nabla/pull/345/files#r1040939380) 

Then we could implement the freeing of all segments in a pool like so
```cpp
void CommandSegmentPool::deleteAll()
{
   CCommandSegment* dummy = nullptr;
   for (CCommandSegment* it=m_firstHeadSegment; it; it=it->header.m_nextHead)
      delete(it->begin(),dummy);
}
```

Reply:
I don't think it will work with a `dummy = nullptr` because, right now, `deleteCommandSegmentList` uses the tail pointer to know where to the end the iteration.